src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.WriteValues.Helpers.cs

```diff
@@ -5,15 +5,118 @@
 using System.Buffers.Text;
 using System.Diagnostics;
 using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
 
 namespace System.Text.Json
 {
     public sealed partial class Utf8JsonWriter
     {
+        private const byte LengthMask =         0b000_000_11;
+        private const byte EncodingMask =       0b000_111_00;
+
+        private const byte Utf8EncodingFlag =   0b000_001_00;
+        private const byte Utf16EncodingFlag =  0b000_010_00;
+
+        private bool TryGetPartialUtf8CodePoint(out ReadOnlySpan<byte> codePointBytes)
+        {
+            ReadOnlySpan<byte> partialCodePointBytes = PartialCodePointRaw;
+            Debug.Assert(partialCodePointBytes.Length == 4);
+
+            if ((partialCodePointBytes[3] & Utf8EncodingFlag) == 0)
+            {
+                codePointBytes = ReadOnlySpan<byte>.Empty;
+                return false;
+            }
+
+            int length = partialCodePointBytes[3] & LengthMask;
+            Debug.Assert((uint)length < 4);
+
+            codePointBytes = partialCodePointBytes.Slice(0, length);
+            return true;
+        }
+
+        private bool TryGetPartialUtf16CodePoint(out ReadOnlySpan<char> codePointChars)
+        {
+            ReadOnlySpan<byte> partialCodePointBytes = PartialCodePointRaw;
+            Debug.Assert(partialCodePointBytes.Length == 4);
+
+            if ((partialCodePointBytes[3] & Utf16EncodingFlag) == 0)
+            {
+                codePointChars = ReadOnlySpan<char>.Empty;
+                return false;
+            }
+
+            int length = partialCodePointBytes[3] & LengthMask;
+            Debug.Assert(length == 2 || length == 0);
+
+            codePointChars = MemoryMarshal.Cast<byte, char>(partialCodePointBytes.Slice(0, length));
+            return true;
+        }
+
+        private void SetPartialUtf8CodePoint(ReadOnlySpan<byte> bytes)
+        {
+            Debug.Assert(bytes.Length <= 3);
+
+            Span<byte> partialCodePointBytes = PartialCodePointRaw;
+
+            bytes.CopyTo(partialCodePointBytes);
+            partialCodePointBytes[3] = (byte)(bytes.Length | Utf8EncodingFlag);
+        }
+
+        private void SetPartialUtf16CodePoint(ReadOnlySpan<char> bytes)
+        {
+            Debug.Assert(bytes.Length <= 1);
+
+            Span<byte> partialCodePointBytes = PartialCodePointRaw;
+
+            bytes.CopyTo(MemoryMarshal.Cast<byte, char>(partialCodePointBytes));
+            partialCodePointBytes[3] = (byte)((2 * bytes.Length) | Utf16EncodingFlag);
+        }
+
+        private bool HasPartialCodePoint => (PartialCodePointRaw[3] & LengthMask) != 0;
+
+        private void ClearPartialCodePoint() => PartialCodePointRaw[3] = 0;
+
+        private void WriteInvalidPartialCodePoint()
```

Comment:
I'm assuming by "invalid" in this case you mean to say that it's incomplete? Isn't that already implied by using "partial"?

Reply:
Looking at the method body I take it "invalid" means that it requires escaping? Is it possible to determine if a a partial code point requires escaping?

Reply:
The caller only calls this when the code is invalid, but what this actually does is write replacement characters for the partial code point. By invalid, I mean that the previous write which set the partial code point was, say, UTF-8 and the new one is UTF-16, so this code point needs to be cleared. Maybe `FlushPartialCodePoint` or similar is a better name (though Utf8JsonWriter.Flush doesn't flush partial code points, so that could be confusing for a different reason).

Reply:
I would probably go with something more verbose, e.g. `WriteReplacementCharForInvalidPartialCodepoint`